micro-rdk/src/esp32/tcp.rs

```diff
@@ -1,129 +1,519 @@
+        let p = unsafe { esp_tls_init() };
+        if p.is_null() {
+            return Err(std::io::Error::new(
+                std::io::ErrorKind::Other,
+                "couldn't allocate tls context",
+            ));
+        }
+        Ok(Self(p))
+    }
+}
+
+impl Deref for Esp32TLSContext {
+    type Target = *mut esp_tls_t;
+    fn deref(&self) -> &Self::Target {
+        &(self.0)
+    }
+}
+
+impl Drop for Esp32TLSContext {
+    fn drop(&mut self) {
+        if let Some(err) = EspError::from(unsafe { esp_tls_conn_destroy(self.0) }) {
+            log::error!("error while dropping the tls connection '{}'", err);
+        }
+    }
+}
+
+#[allow(dead_code)]
+struct Esp32ServerConfig {
+    cfg: Box<esp_tls_cfg_server>,
+    alpn_proto: Vec<*const c_char>,
+}
+
+impl Esp32ServerConfig {
+    fn new(srv_cert: &[u8], srv_key: &[u8]) -> Self {
+        let mut alpn_proto: Vec<_> = vec![ALPN_PROTOCOLS.as_ptr() as *const i8, std::ptr::null()];
+        let cfg = Box::new(esp_tls_cfg_server {
+            alpn_protos: alpn_proto.as_mut_ptr(),
+            __bindgen_anon_1: crate::esp32::esp_idf_svc::sys::esp_tls_cfg_server__bindgen_ty_1 {
+                // The CA root is not need when a client is connecting as it's available
+                cacert_buf: std::ptr::null(),
+            },
+            __bindgen_anon_2: crate::esp32::esp_idf_svc::sys::esp_tls_cfg_server__bindgen_ty_2 {
+                cacert_bytes: 0,
+            },
+            __bindgen_anon_3: crate::esp32::esp_idf_svc::sys::esp_tls_cfg_server__bindgen_ty_3 {
+                // This is the server certificates in the PEM format
+                servercert_buf: srv_cert.as_ptr(),
+            },
+            __bindgen_anon_4: crate::esp32::esp_idf_svc::sys::esp_tls_cfg_server__bindgen_ty_4 {
+                servercert_bytes: srv_cert.len() as u32,
+            },
+            __bindgen_anon_5: crate::esp32::esp_idf_svc::sys::esp_tls_cfg_server__bindgen_ty_5 {
+                serverkey_buf: srv_key.as_ptr(),
+            },
+            __bindgen_anon_6: crate::esp32::esp_idf_svc::sys::esp_tls_cfg_server__bindgen_ty_6 {
+                serverkey_bytes: srv_key.len() as u32,
+            },
+            serverkey_password: std::ptr::null(),
+            serverkey_password_len: 0_u32,
+        });
+        Self { cfg, alpn_proto }
+    }
+    fn get_cfg_ptr(&self) -> *const esp_tls_cfg_server {
+        &*self.cfg as *const _
+    }
+}
+
+#[allow(dead_code)]
+struct Esp32ClientConfig {
+    cfg: Box<esp_tls_cfg>,
+    alpn_proto: Vec<*const c_char>,
+}
+
+impl Esp32ClientConfig {
+    fn new() -> Self {
+        let mut alpn_proto = vec![ALPN_PROTOCOLS.as_ptr() as *const i8, std::ptr::null()];
+        // this is a root certificate to validate the server's certificate
+        let cert = include_bytes!("../../certs/google_gts_root_r1.crt");
+
+        let cfg = Box::new(esp_tls_cfg {
+            alpn_protos: alpn_proto.as_mut_ptr(),
```

Comment:
does this need to be pinned or copied? I think the bug @acmorrow came across when adding http2 support was related to something like this or like what's below.